cpp/daal/src/algorithms/kmeans/kmeans_plusplus_init_impl.i

```diff
@@ -287,20 +287,29 @@ public:
         return _csr->releaseSparseBlock(block);
     }
 
+    //calculate distance from current trial center to the rows in the block and update min distance
     algorithmFPType updateMinDistForITrials(algorithmFPType * const pDistSq, size_t iTrials, size_t nRowsToProcess,
                                             const algorithmFPType * const pData, const size_t * const colIdx, const size_t * const rowIdx,
                                             const algorithmFPType * const pLastAddedCenter, const algorithmFPType * const aWeights,
                                             const algorithmFPType * const pDistSqBest)
     {
-        algorithmFPType sumOfDist2 = algorithmFPType(0);
-        size_t csrCursor           = 0u;
+        algorithmFPType sumOfDist2            = algorithmFPType(0);
+        size_t csrCursor                      = 0u;
+        algorithmFPType pLastAddedCenterSumSq = algorithmFPType(0.);
+        for (size_t iCol = 0u; iCol < dim; iCol++)
+        {
+            pLastAddedCenterSumSq += pLastAddedCenter[iCol] * pLastAddedCenter[iCol];
+        }
```

Comment:
This was the computation previously:
const size_t nValues  = rowIdx[iRow + 1] - rowIdx[iRow];
for (size_t i = 0u; i < nValues; i++, csrCursor++)
            {
                dist2 += (pData[csrCursor] - pLastAddedCenter[colIdx[csrCursor] - 1])^2
            }

For a row in pData [1, 0, 0] and pLastAddedCenter = [0, 0, 1], the distance with previous scheme will be 1. Because nValues = 1, so pData[0] = 1 and pLastAddedCenter[0] = 0. But the correct distance should be 2. The current scheme does not take into account that pLastAddedCenter can have non zero elements at indices where pData has zero. The proposed scheme fixes this by expanding (a-b)^2 to a^2 + b^2 -2ab.

Reply:
Yes, I understand that. I meant that it would be good to describe in the comments in the code.

Reply:
my bad :), I will update the code